File: examples/pv_stats.py
"""Collect N samples of a PV at a fixed rate and compute statistics.

No loop. No counter. No list management. The Rx chain handles all of it:
  interval → take(N) → read each tick → collect → reduce to stats

Usage:
    python pv_stats.py <pv_name> [samples] [interval-ms]

Examples:
    python pv_stats.py TEST:CALC
    python pv_stats.py TEST:CALC 50 500
"""

import asyncio
import sys
from datetime import timedelta
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

import reactivex as rx
import reactivex.operators as ops
from reactivex.scheduler.eventloop import AsyncIOScheduler
from caproto.asyncio.client import Context

from rxepics.channel import read_pv


async def main():
    if len(sys.argv) < 2:
        print("Usage: pv_stats.py <pv_name> [samples] [interval-ms]", file=sys.stderr)
        sys.exit(1)

    pv_name = sys.argv[1]
    n = int(sys.argv[2]) if len(sys.argv) > 2 else 20
    interval_ms = int(sys.argv[3]) if len(sys.argv) > 3 else 500

    loop = asyncio.get_running_loop()
    scheduler = AsyncIOScheduler(loop)
    ctx = Context()

    print(f"Collecting {n} samples of {pv_name} @ {interval_ms} ms intervals...\n")

    counter = 0
    samples: list[float] = []
    done = asyncio.Event()

    def on_next(v: float) -> None:
        nonlocal counter
        counter += 1
        samples.append(v)
        print(f"  [{counter:2d}]  {v:+.6f}")

    def on_completed() -> None:
        if not samples:
            done.set()
            return
        n_actual = len(samples)
        mean = sum(samples) / n_actual
        variance = sum((x - mean) ** 2 for x in samples) / n_actual
        import math
        print(f"\n  n      = {n_actual}")
        print(f"  min    = {min(samples):+.6f}")
        print(f"  max    = {max(samples):+.6f}")
        print(f"  mean   = {mean:+.6f}")
        print(f"  stddev = {math.sqrt(variance):.6f}")
        done.set()

    # interval + take(N): emit exactly N ticks, then complete automatically.
    # flat_map: one read per tick.
    # to_list: Rx handles list allocation and synchronisation.
    rx.interval(timedelta(milliseconds=interval_ms), scheduler=scheduler).pipe(
        ops.take(n),
        ops.flat_map(lambda _: read_pv(pv_name, ctx)),
        ops.do_action(on_next=on_next),
        ops.to_list(),
    ).subscribe(
        on_next=lambda _: None,  # data already consumed by do_action
        on_error=lambda e: (print(f"ERROR: {e}", file=sys.stderr), done.set()),
        on_completed=on_completed,
        scheduler=scheduler,
    )

    await done.wait()


if __name__ == "__main__":
    asyncio.run(main())

File: RxEpics/cpp/examples/pv_stats.cpp
/**
 * Collect N samples from a PV and print batch statistics, then exit.
 *
 * Mirrors Python's pv_stats.py.
 *
 * Usage:
 *   EPICS_PVA_ADDR_LIST=localhost ./pv_stats [pv] [N] [interval-ms]
 *   defaults: TEST:CALC  20  200
 */

#include <algorithm>
#include <chrono>
#include <condition_variable>
#include <cmath>
#include <iostream>
#include <mutex>
#include <numeric>
#include <string>
#include <vector>

#include <rxcpp/rx.hpp>
#include <rxepics/rxepics.hpp>

int main(int argc, char* argv[]) {
    const std::string pv          = argc > 1 ? argv[1] : "TEST:CALC";
    const int         N           = argc > 2 ? std::stoi(argv[2]) : 20;
    const int         interval_ms = argc > 3 ? std::stoi(argv[3]) : 200;

    std::cout << "Collecting " << N << " samples of " << pv << " ...\n\n";

    auto& ctx = rxepics::default_context();

    std::mutex              m;
    std::condition_variable cv;
    bool                    done = false;

    rxcpp::observable<>::interval(std::chrono::milliseconds(interval_ms))
        .flat_map([pv, &ctx](long) { return rxepics::read_pv<double>(pv, ctx); })
        .take(N)
        .to_vector()
        .subscribe(
            [N](std::vector<double> samples) {
                double sum  = std::accumulate(samples.begin(), samples.end(), 0.0);
                double mean = sum / N;
                double var  = 0.0;
                for (double x : samples) var += (x - mean) * (x - mean);
                var /= (N - 1);

                std::cout << "  N      = " << N << "\n"
                          << "  min    = " << *std::min_element(samples.begin(), samples.end()) << "\n"
                          << "  max    = " << *std::max_element(samples.begin(), samples.end()) << "\n"
                          << "  mean   = " << mean << "\n"
                          << "  stddev = " << std::sqrt(var) << "\n";
            },
            [&](std::exception_ptr e) {
                try { std::rethrow_exception(e); }
                catch (std::exception& ex) { std::cerr << "  ERROR: " << ex.what() << "\n"; }
                std::lock_guard<std::mutex> lk(m); done = true; cv.notify_one();
            },
            [&]() { std::lock_guard<std::mutex> lk(m); done = true; cv.notify_one(); }
        );

    std::unique_lock<std::mutex> lk(m);
    cv.wait(lk, [&]{ return done; });
    return 0;
}
